Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp

```diff
@@ -29,118 +29,266 @@
 
 #include "ClipPath.h"
+#include "FloatBox3D.h"
 #include "FloatPlane3D.h"
+#include "FloatPolygon3D.h"
+#include "GeometryUtilities.h"
 #include "TextureMapperGPUBuffer.h"
 #include "TextureMapperLayer.h"
-#include <wtf/Deque.h>
+#include <numeric>
+#include <wtf/StdSet.h>
 #include <wtf/TZoneMallocInlines.h>
 
 namespace WebCore {
 
 WTF_MAKE_TZONE_ALLOCATED_IMPL(TextureMapperLayer3DRenderingContext);
 
-void TextureMapperLayer3DRenderingContext::paint(TextureMapper& textureMapper, const Vector<TextureMapperLayer*>& layers,
+void TextureMapperLayer3DRenderingContext::paint(TextureMapper& textureMapper, const Vector<TextureMapperLayer*>& textureMapperLayers,
     const std::function<void(TextureMapperLayer*, const ClipPath&)>& paintLayerFunction)
 {
-    if (layers.isEmpty())
+    if (textureMapperLayers.isEmpty())
         return;
 
-    Deque<TextureMapperLayerPolygon> layerList;
-    for (auto* layer : layers)
-        layerList.append({ { layer->effectiveLayerRect(), layer->toSurfaceTransform() }, layer, false });
-
-    auto root = makeUnique<TextureMapperLayerNode>(layerList.takeFirst());
-    buildTree(*root, layerList);
-
-    // Collect clip data
-    Vector<float> clipVertices;
-    traverseTree(*root, [&clipVertices](TextureMapperLayerNode& node) {
-        for (auto& polygon : node.polygons) {
-            auto toLayerTransform = polygon.layer->toSurfaceTransform().inverse();
-            if (polygon.isSplitted && toLayerTransform) {
-                polygon.clipVertexBufferOffset = clipVertices.size();
-
-                unsigned numVertices = polygon.geometry.numberOfVertices();
-                for (unsigned i = 0; i < numVertices; i++) {
-                    auto v = toLayerTransform->mapPoint(polygon.geometry.vertexAt(i));
-                    clipVertices.append(v.x());
-                    clipVertices.append(v.y());
-                }
+    Vector<Layer> layers;
+    for (auto* layer : textureMapperLayers)
+        layers.append({ { layer->effectiveLayerRect(), layer->toSurfaceTransform() }, layer, false });
+
+    // Perform a broad-phase sweep-and-prune across x axis to identify potential intersections.
+    // By determining which layers might intersect, we can limit BSP cutting planes to those areas only,
+    // preventing unnecessary splitting of layers that are spatially distant.
+    // This optimization matters because the TextureMapper uses stencil operations to render split layers,
+    // and on some hardware, stencil usage is slow. Reducing unnecessary splits helps improve performance.
+    auto potentialIntersections = sweepAndPruneXAxis(layers);
+
+    // Determine if any layer pairs intersect on the ZY plane. An intersection implies that the rendering
+    // order is either ambiguous or overlapping, necessitating the use of a BSP tree for proper ordering.
+    bool hasIntersections = false;
+    if (!potentialIntersections.isEmpty()) {
+        for (const auto& p : potentialIntersections) {
```

Comment:
I don't think we need to check if it's empty, the loop won't iterate anyway.

Reply:
Fixed